Mathlib/RingTheory/Extension.lean

```diff
@@ -0,0 +1,389 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.LinearAlgebra.TensorProduct.RightExactness
+import Mathlib.RingTheory.Ideal.Cotangent
+
+/-!
+
+# Extension of algebras
+
+## Main definition
+
+- `Algebra.Extension`: An extension of an `R`-algebra `S` is an `R` algebra `P` together with a
+surjection `P →ₐ[R] R`.
+
+- `Algebra.Extension.Hom`: Given a commuting square
+  ```
+  R --→ P -→ S
+  |          |
+  ↓          ↓
+  R' -→ P' → S
+  ```
+  A hom between `P` and `P'` is a ring homomorphism that makes the two squares commute.
+
+- `Algebra.Extension.Cotangent`:
+  The cotangent space wrt an extension `P → S` by `I`, i.e. the space `I/I²`.
+
+-/
+
+universe w u v
+
+open TensorProduct MvPolynomial
+
+variable (R : Type u) (S : Type v) [CommRing R] [CommRing S] [Algebra R S]
+
+/--
+An extension of an `R`-algebra `S` is an `R` algebra `P` together with a surjection `P →ₐ[R] S`.
+Also see `Algebra.Extension.ofSurjective`.
+-/
+structure Algebra.Extension where
+  /-- The underlying algebra of an extension. -/
+  Ring : Type w
+  [commRing : CommRing Ring]
+  [algebra₁ : Algebra R Ring]
+  [algebra₂ : Algebra Ring S]
+  [isScalarTower : IsScalarTower R Ring S]
+  /-- A chosen (set-theoretic) section of an extension. -/
+  σ : S → Ring
```

Comment:
Doesn't this mean that two extensions that should be equal (or isomorphic) will not be so, because they might have non-isomorphic sections?

Are there places where you require good defeqs of this section?

Reply:
I think under our definition of morphisms, isomorphism doesn't care about the sections.
I don't think we require defeqs but we had them in presentations and we will lose them if we drop them here.